File: java/AtomNone.java
/*
 *  Copyright (c) 2014-2015 Bruce Twite
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Special Atom to signify that we have no argument or value.
 */
package yaga.core;

/**
 *
 * @author Bruce
 */
public class AtomNone extends AtomicList
{
	static public AtomNone VALUE = new AtomNone();
			
	private AtomNone()
		{ super(ParserPoint.Default); }
	
	@Override
	public AtomNone zeroValue()
		{ return (this); }
	
	@Override
	public boolean isNone()
		{ return (true); }

	@Override
	public void print(StringBuilder sb) 
		{ Symbol.printReserved(sb, Symbol.Reserved.NONE); }
	
	@Override
	public List neg(Context ctxt)
		{ return (this); }
	@Override
	public List add(Context ctxt, List e)
		{ return (this); }
	@Override
	public List sub(Context ctxt, List e)
		{ return (this); }
	@Override
	public List mul(Context ctxt, List e)
		{ return (this); }
	@Override
	public List div(Context ctxt, List e)
		{ return (this); }
	@Override
	public List rem(Context ctxt, List e)
		{ return (this); }
	
	@Override
	public List rAdd(Context ctxt, List e)
		{ return (this); }
	@Override
	public List rSub(Context ctxt, List e)
		{ return (this); }
	@Override
	public List rMul(Context ctxt, List e)
		{ return (this); }
	@Override
	public List rDiv(Context ctxt, List e)
		{ return (this); }
	@Override
	public List rRem(Context ctxt, List e)
		{ return (this); }
}

File: AtomNone.js
/*
 *  AtomNone : @file
 *
 *  Special Atom to signify that we have no argument or value.
 */
"use strict";

var yc, _prot;
module.exports = {
	new: _newAtomNone,
	VALUE: undefined,
	Initialise: _init,
};

function _init(y) {
	if (yc) return;
	yc = y;
	_prot = Object.assign(Object.create(yc.AtomicList.prototype, {
		zeroValue() {
			return (this);
		},
		isNone: () => true,
		print(sb) {
			yc.Symbol.printReserved(sb, yc.Symbol.Reserved.NONE);
		},
		neg(ctxt) {
			return (this);
		},
		add(ctxt, e) {
			return (this);
		},
		sub(ctxt, e) {
			return (this);
		},
		mul(ctxt, e) {
			return (this);
		},
		div(ctxt, e) {
			return (this);
		},
		rem(ctxt, e) {
			return (this);
		},
		rAdd(ctxt, e) {
			return (this);
		},
		rSub(ctxt, e) {
			return (this);
		},
		rMul(ctxt, e) {
			return (this);
		},
		rDiv(ctxt, e) {
			return (this);
		},
		rRem(ctxt, e) {
			return (this);
		},
	}));
	module.exports.VALUE = _newAtomNone();
	Object.freeze(module.exports);
}

function _newAtomNone() {
	return (Object.create(_prot));
}